magma/t.py

```diff
@@ -164,6 +164,10 @@ def flip(cls):
     def undirected_t(cls):
         return cls.qualify(Direction.Undirected)
 
+    @property
```

Comment:
think we can just do `return cls.direction is not None`?

Reply:
Not quite, I think only leaf (Digital) types define the `direction` attribute.  The default implementation of flip is misleading, since arrays/tuples define their own implementations of flip (so it's not really inherited and probably should be removed).

For example, a tuple doesn't necessarily have a direction (it could be mixed based on the children), so I think the most general thing to check is whether the undirected version of the type is equal to the current object.  Alternatively we could have each class implement it specifically (probably more efficient, since you can check if any of the children are directed for tuple and short circuit if you find one), but this avoids having to define an implementation for each type (by using the qualify API).